### PR TITLE
allow user to specify task metadata from client

### PR DESCRIPTION
**Pull request by rhyw**
Related: https://issues.redhat.com/browse/OSH-567
Closes: https://github.com/openscanhub/openscanhub/pull/285

### CODE REVIEW

**Comment by rhyw** [comment]
So the original request asks for the option for `mock-build`, I think it probably makes more sense to make it available for  `mock-build`, `diff-build`, `version-diff-build`, thus the patch.

We don't have filtering capabilities for task args for `find-tasks` subcommand. I had some local test, it supports options like `args__contains`, to filter values for a specific key in `args` is not supported right now.

**Review by kdudka — APPROVED** [review]
I have not tested it myself but at first glance it looks sane.

**Comment by rhyw** [comment]
added one more commit such that `--metadata` is available in command completion.

File: osh/client/commands/common.py [hunk 1]
```diff
@@ -150,3 +154,24 @@ def add_tarball_option(parser):
              "This option sets command which should build the package, usually this should be just "
              "\"make\", in case of packages which doesn't need to be built, just pass \"true\".",
     )
+
+
+def parse_json_option(option, opt, value, parser):
+    try:
+        json_value = json.loads(value)
+    except json.JSONDecodeError as e:
+        raise OptionValueError(f"Option {opt}: invalid JSON value: {e}")
```
Comment (lzaoral) [inline osh/client/commands/common.py:163]: This is a good idea but the error message looks a bit too convoluted:
```console
$ osh-cli mock-build --metadata=
Usage: osh-cli mock-build [options] <args>

osh-cli: error: Option --metadata: invalid JSON value: Expecting value: line 1 column 1 (char 0)
```

File: osh/hub/scan/check.py [hunk 2]
```diff
@@ -121,3 +122,15 @@ def check_srpm(upload_id, build_nvr, task_user, is_tarball=False):
...
+
+def check_task_metadata(metadata):
+    if isinstance(metadata, dict):
+        return metadata
+    elif isinstance(metadata, str):
+        try:
+            return json.loads(metadata)
+        except json.JSONDecodeError as e:
+            raise RuntimeError(f"Invalid JSON value: {e}")
+    else:
```
Comment (lzaoral) [inline osh/hub/scan/check.py:135]: This `else` is redundant because all previous blocks end with a jump statement.

File: osh/hub/scan/check.py [hunk 3]
```diff
@@ -121,3 +122,15 @@ def check_srpm(upload_id, build_nvr, task_user, is_tarball=False):
         }
 
     raise RuntimeError('No source RPM or tarball specified.')
+
+
+def check_task_metadata(metadata):
+    if isinstance(metadata, dict):
+        return metadata
+    elif isinstance(metadata, str):
```
Comment (lzaoral) [inline osh/hub/scan/check.py:130]: This can be a plain `if` because all previous blocks end with a jump statement.

File: osh/hub/scan/check.py [hunk 4]
```diff
@@ -121,3 +122,15 @@ def check_srpm(upload_id, build_nvr, task_user, is_tarball=False):
...
+        try:
+            return json.loads(metadata)
+        except json.JSONDecodeError as e:
+            raise RuntimeError(f"Invalid JSON value: {e}")
```
Comment (lzaoral) [inline osh/hub/scan/check.py:134]: You can simplify this to `RuntimeError("Invalid JSON value:", e)`

File: osh/hub/scan/scanner.py [hunk 5]
```diff
@@ -487,6 +488,8 @@ def validate_options(self):
         self.client_csmock_args = self.options.get('csmock_args', None)
 
         self.metadata = self.options.get('metadata')
+        if self.metadata:
+            self.metadata = check_task_metadata(self.metadata)
```
Comment (lzaoral) [inline osh/hub/scan/scanner.py]: nit: You can move this check to `check_task_metadata` as well.

**Review by lzaoral — CHANGES_REQUESTED** [review]
Thank you, @rhyw!  I have tested the feature and it seems to work as expected!  Some comments inline.

File: osh/client/commands/common.py (hunk 1 above)
Comment (rhyw) [reply osh/client/commands/common.py]: I've updated the exception to print the original json string instead. Thanks for suggestion!

File: osh/client/commands/common.py [hunk 6]
```diff
@@ -150,3 +154,24 @@ def add_tarball_option(parser):
...
+
+    setattr(parser.values, option.dest, json_value)
+
+
+def add_task_metadata_option(parser):
+    parser.add_option(
+        "--metadata",
+        dest="metadata",
+        type="string",
```
[the remark this answers is not in the document]
Comment (rhyw) [reply osh/client/commands/common.py]: This line's removed. I think it's ok to keep it too.

File: osh/client/completion/_osh-cli [hunk 7]
```diff
@@ -125,6 +125,7 @@ _osh-cli_args()
                 '--config=[specify mock config name]:osh:->configs'
                 '--comment=[a task description]'
                 '--json[print created task info in JSON]'
+                '--metadata[specify task metadata as a JSON string]'
```
[the remark this answers is not in the document]
Comment (rhyw) [reply osh/client/completion/_osh-cli]: Good catch, this indeed is an issue, fixed.

File: osh/hub/scan/scanner.py (hunk 5 above)
Comment (rhyw) [reply osh/hub/scan/scanner.py]: This is reasonable in some way. I'd like to keep it as it as so the `check_task_metadata` only deals with non-null values.

**Comment by rhyw** [comment]
@lzaoral Thanks for testing. Patch updated per suggestions.

**Review by lzaoral — APPROVED** [review]

**Comment by rhyw** [comment]
Thanks for the review!

**Pull request merged**